tests/test_auth.py

```diff
@@ -227,3 +227,19 @@ def test_warns_for_empty_password(
     assert auth.Resolver(config, auth.CredentialInput()).password == password
 
     assert caplog.messages[0].startswith(warning)
+
+
+def test_log_exception_on_keyring_failure(config, monkeypatch, caplog):
+    class FailKeyring:
+        @staticmethod
+        def get_credential(system, username):
+            # Simulate the error from https://github.com/pypa/twine/issues/889
+            environ = {}
+            environ["HOME"]
+
+    monkeypatch.setattr(auth, "keyring", FailKeyring())
+
+    assert not auth.Resolver(config, auth.CredentialInput()).get_username_from_keyring()
+    assert not auth.Resolver(config, auth.CredentialInput()).get_password_from_keyring()
+
+    assert caplog.messages[0] == "Error from keyring" and "KeyError: 'HOME'" in caplog.text
```

Comment:
I think a regular expression would make this clearer to future maintainers:

```suggestion
    assert re.search(
        r"Error from keyring.+Traceback.+KeyError: 'HOME'", caplog.text, re.DOTALL
    )
```

Same suggestion applies to the tests above.